src/ast/recfun_decl_plugin.cpp

```diff
@@ -0,0 +1,434 @@
+/*++
+Module Name:
+
+    recfun_decl_plugin.cpp
+
+Abstract:
+
+    Declaration and definition of (potentially recursive) functions
+
+Author:
+
+    Simon Cruanes 2017-11
+
+Revision History:
+
+--*/
+
+#include <sstream>
+#include "ast/expr_functors.h"
+#include "ast/expr_substitution.h"
+#include "ast/recfun_decl_plugin.h"
+#include "ast/ast_printer.h"
+
+namespace recfun {
+
+    void case_pred::set_name(std::string & s) {
+        m_name_buf = std::string(s); // owned copy
+        m_name = symbol(m_name_buf.c_str());
+    }
+
+    case_def::case_def(ast_manager &m,
+                       def * d,
+                       std::string & name,
+                       sort_ref_vector const & arg_sorts,
+                       expr* guards, unsigned num_guards, expr* rhs)
+    : m_pred(arg_sorts), m_guards(), m_rhs(expr_ref(rhs,m)), m_def(d) {
+        m_pred.set_name(name);
+        for (unsigned i=0; i<num_guards; ++i) {
+            m_guards.push_back(expr_ref(&guards[i], m));
+        }
+    }
+
+    def::def(ast_manager &m, symbol const & s, sort * args, unsigned n_args, sort* ret)
+        :   m_kind(OP_FUN_DEFINED), m_manager(m), m_name(s),
+            m_arg_sorts(m), m_ret_sort(ret, m), m_vars(), m_cases()
+    {
+        for (unsigned i=0; i<n_args; ++i) m_arg_sorts.push_back(sort_ref(&args[i], m));
+    }
+
+    // does `e` contain any `ite` construct?
+    bool def::contains_ite(expr * e) {
+        struct ite_find_p : public i_expr_pred {
+            ast_manager & m;
+            ite_find_p(ast_manager & m) : m(m) {}
+            virtual bool operator()(expr * e) { return m.is_ite(e); }
+        };
+        ite_find_p p(m());
+        check_pred cp(p, m());
+        return cp(e);
+    }
+
+    /*
+     * compilation of functions to a list of cases
+     */
+    
+    class compute_case_op; // fwd decl of operations called during exploration of cases
+    class compute_case_alternative; // fwd decl of operations to explore alternative cases
+
+    // state for compiling a function's definition into cases
+    struct compute_case_state {
+        ast_manager &               m_manager;
+        region                      m_region;
```

Comment:
Do you really need a region inside of here?
It is mostly useful in settings where a lot of memory is allocated and then reclaimed quickly, such as during backtracking.